### DISCUSSION TITLE
Bit Array

### DISCUSSION THREAD

**Discussion by onee-only** [subak-ba/cs-study#46] [자료 구조]
# Bit Array
비트 배열(bit array, 또는 bitmask, bit set, bit map, bit string, bit vector 등)은 비트를 보관하는 배열 형태의 자료 구조입니다.
저장 단위가 비트이기 때문에 다른 자료 구조보다 훨씬 공간을 적게 사용합니다.

w가 저장되는 원소의 비트 수이고, k가 그 원소의 개수일 때, 비트 배열은 k * w 비트를 사용합니다.
w는 일반적으로 바이트, 워드의 비트 수를 사용합니다. 

비트 배열은 Set를 구현할 수 있으며, [Bloom Filter](https://github.com/subak-ba/cs-study/discussions/32)의 기반 자료 구조로 사용됩니다.

## 비트 배열의 장점
- [비트 수준 병렬 처리](https://en.wikipedia.org/wiki/Bit-level_parallelism)를 효과적으로 이용할 수 있습니다.
- 배열 순회의 **최대** 메모리 접근 횟수가 `n / w`로 다른 자료 구조보다 훨씬 적습니다.
- 캐시 라인의 워드 수를 k라고 했을 때, `n / (w * k)`번의 캐시 미스가 생깁니다.

## 구현
Go 언어로 비트 배열을 구현했습니다.
```go
package main

import "fmt"

// word size
//
// 64비트일 때: (^uint(0) >> 63) == 1
// w = 64 = 32 << 1
//
// 32비트일 때: (^uint(0) >> 63) == 0
// w = 32 = 32 << 0
const w = 32 << (^uint(0) >> 63)

type bitarray struct {
	arr []uint // word array
}

// idx의 워드에서의 비트 위치를 가져와 그 비트를 1로 설정한 값을 반환
func mask(idx int) uint { return uint(1 << (idx % w)) }

func (arr *bitarray) check(idx int) bool {
	mask := mask(idx)

	word := arr.arr[idx/w]
	return word&mask != 0
}

func (arr *bitarray) set(idx int, b bool) {
	mask := mask(idx)
	if b {
		arr.arr[idx/w] |= mask
	} else {
		arr.arr[idx/w] &= ^mask
	}
}

func (arr *bitarray) grow(words int) {
	if words <= len(arr.arr) {
		return
	}

	newarr := make([]uint, words)

	// for idx, word := range arr.arr {
	// 	newarr[idx] = word
	// }
	copy(newarr, arr.arr) // == 위 for문

	arr.arr = newarr
}

func main() {
	barr := new(bitarray)
	barr.grow(2) // 2 words

	barr.set(3, true)
	barr.set(4, true)
	barr.set(5, true)

	fmt.Println(barr.check(2) == false)
	fmt.Println(barr.check(3) == true)
	fmt.Println(barr.check(4) == true)
	fmt.Println(barr.check(5) == true)
	fmt.Println(barr.check(6) == false)

	barr.set(4, false)
	fmt.Println(barr.check(4) == false)
}
```
java의 구현체는 [이곳에서](https://github.com/openjdk-mirror/jdk7u-jdk/blob/master/src/share/classes/java/util/BitSet.java) 볼 수 있습니다.

## 문제
[중복 제거](https://www.acmicpc.net/problem/13701) (골드 4)

**Comment by onee-only** [comment]
문제 해결은 직접 구현보다는 java 구현체를 사용하는 것을 추천드려요!

**Comment by BLoHny** [comment]
https://github.com/BLoHny/KOTE/blob/main/%EB%B0%B1%EC%A4%80/Gold/13701.%E2%80%85%EC%A4%91%EB%B3%B5%E2%80%85%EC%A0%9C%EA%B1%B0/%EC%A4%91%EB%B3%B5%E2%80%85%EC%A0%9C%EA%B1%B0.java

문제 풀었습니다!
